homeassistant/components/tplink/camera.py

```diff
@@ -0,0 +1,156 @@
+):
+    """Base class for camera entity description."""
+
+
+CAMERA_DESCRIPTIONS: tuple[TPLinkCameraEntityDescription, ...] = (
+    TPLinkCameraEntityDescription(
+        key="live_view",
+        translation_key="live_view",
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: TPLinkConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up camera entities."""
+    data = config_entry.runtime_data
+    parent_coordinator = data.parent_coordinator
+    device = parent_coordinator.device
+    ffmpeg_manager = ffmpeg.get_ffmpeg_manager(hass)
+
+    async_add_entities(
+        TPLinkCameraEntity(
+            device,
+            parent_coordinator,
+            description,
+            camera_module=camera_module,
+            parent=None,
+            ffmpeg_manager=ffmpeg_manager,
+        )
+        for description in CAMERA_DESCRIPTIONS
+        if (camera_module := device.modules.get(Module.Camera))
+    )
+
+
+class TPLinkCameraEntity(CoordinatedTPLinkEntity, Camera):
+    """Representation of a TPLink camera."""
+
+    _attr_supported_features = CameraEntityFeature.STREAM | CameraEntityFeature.ON_OFF
+
+    entity_description: TPLinkCameraEntityDescription
+
+    def __init__(
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        description: TPLinkCameraEntityDescription,
+        *,
+        camera_module: CameraModule,
+        parent: Device | None = None,
+        ffmpeg_manager: ffmpeg.FFmpegManager,
+    ) -> None:
+        """Initialize a TPlink camera."""
+        self.entity_description = description
+        self._camera_module = camera_module
+        self._video_url: str | None = None
+        self._image: bytes | None = None
+        super().__init__(device, coordinator, parent=parent)
+        Camera.__init__(self)
+        self._ffmpeg_manager = ffmpeg_manager
+        self._image_lock = asyncio.Lock()
+
+    def _get_unique_id(self) -> str:
+        """Return unique ID for the entity."""
+        return f"{legacy_device_id(self._device)}-{self.entity_description}"
+
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Update the entity's attributes."""
+        self._attr_is_on = self._camera_module.is_on
+        self._video_url = self._camera_module.stream_rtsp_url()
+
+    async def stream_source(self) -> str | None:
+        """Return the source of the stream."""
+        return self._video_url
+
+    async def async_camera_image(
+        self, width: int | None = None, height: int | None = None
+    ) -> bytes | None:
+        """Return a still image response from the camera."""
+        if self._image is None and (video_url := self._video_url):
+            # Sometimes the front end makes multiple image requests
+            async with self._image_lock:
+                if self._image:
+                    return self._image  # type: ignore[unreachable]
```

Comment:
I feel a bit uncomfortable about this structure, are other camera platforms using similar?

Reply:
`doorbird` and `proxy` are using a last update time check, `push` is using a queue, `amcrest` waits on a previous async task to finish, and `buienradar` uses an `asyncio.Condition`

Reply:
I'm all for de-duping the requests, but isn't this going to cache the first request forever?

Reply:
So this will now refresh the image on a 5 minute interval